tools/schemapi/schemapi.py

```diff
@@ -1120,17 +1366,16 @@ def from_dict(
         """
         if validate:
             cls.validate(dct)
-        converter = _FromDict(cls._default_wrapper_classes())
+        converter: type[_FromDict] | _FromDict = (
+            _FromDict
+            if _FromDict.hash_tps
+            else _FromDict(cls._default_wrapper_classes())
+        )
         return converter.from_dict(dct, cls)
```

Comment:
NOTE: the breadth-first search now occurs only once.

`_FromDict` is purely `ClassVar`/`@classmethod`(s)